src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs

```diff
@@ -193,10 +160,23 @@ public static Quaternion Identity
                 var left = value1.AsVector128();
                 var right = value2.AsVector128();
 
-                var result = right * left.GetElementUnsafe(3);
-                result += (Vector128.Shuffle(right, Vector128.Create(3, 2, 1, 0)) * left.GetElementUnsafe(0)) * Vector128.Create(+1.0f, -1.0f, +1.0f, -1.0f);
-                result += (Vector128.Shuffle(right, Vector128.Create(2, 3, 0, 1)) * left.GetElementUnsafe(1)) * Vector128.Create(+1.0f, +1.0f, -1.0f, -1.0f);
-                result += (Vector128.Shuffle(right, Vector128.Create(1, 0, 3, 2)) * left.GetElementUnsafe(2)) * Vector128.Create(-1.0f, +1.0f, +1.0f, -1.0f);
+                var l0012 = Vector128.Shuffle(left, Vector128.Create(0, 0, 1, 2));
+                var l1120 = Vector128.Shuffle(left, Vector128.Create(1, 1, 2, 0));
+                var r0333 = Vector128.Shuffle(right, Vector128.Create(0, 3, 3, 3));
+                var r1201 = Vector128.Shuffle(right, Vector128.Create(1, 2, 0, 1));
+
+                var t0 = l0012 * r0333 + l1120 * r1201;
+                var mask = Vector128.Create(0x80000000, 0, 0, 0);
+                var t0m = Vector128.Xor(t0, mask.AsSingle());
+
+                var l2201 = Vector128.Shuffle(left, Vector128.Create(2, 2, 0, 1));
+                var l3333 = Vector128.Shuffle(left, Vector128.Create(3, 3, 3, 3));
+                var r2120 = Vector128.Shuffle(right, Vector128.Create(2, 1, 2, 0));
+                var r3012 = Vector128.Shuffle(right, Vector128.Create(3, 0, 1, 2));
+
+                var t1 = l3333 * r3012 - l2201 * r2120 + t0m;
+                var result = Vector128.Shuffle(t1, Vector128.Create(1, 2, 3, 0));
```

Comment:
This new code is much harder to read for basically a cycle difference in perf (numbers above report `0.7 -> 0.5 ns`). It may also not optimize as well when considering all platforms or ISA baselines (SSE2, SSE4.1, AVX2, AVX512F, AdvSimd, WASM, etc).

It's also substantially different from `Concatenate` despite them doing the "same thing" (just one is `x * y` and the other is `y * x`)

Reply:
It's an operation that tends to be in the hot path for 3d applications so thought the reduced readability might be worth the extra performance on all the platforms I tested it on.

Reply:
Its not that hot for 3D apps. The actual hot code tends to happen on the GPU instead.

Its especially not so hot that the complexity is worth `0.15ns` (which is effectively 1 CPU cycle).

Reply:
This comment is still pending. I don't think the new code is worth the 0.15ns savings (it's notably slightly slower on my box too). The code is more complex, harder to follow, and so I don't think the justification is there to take this part of the change.

.NET is huge, it is millions of lines of code. We ultimately have to balance readability, maintainability, security, perf, etc.

While `Vector3` is itself a performance oriented type, we still have to factor in the other considerations and balance them out. It also applies beyond what a simple benchmark may measure, as we have multiple platforms, ISAs, and ABIs to consider. We also have to consider how certain optimizations may function (such as inlining and constant folding).

As such, even with it being a performance oriented type, there's many places where we are happy to give up a half nanosecond (which is in practice 1-5 instruction cycles) to improve the other areas.